### PR TITLE
Add database restore capability

### PR DESCRIPTION
**Pull request by merrilmathew**
### Context

Postgres database can be restored from point in time backups available in Gov PaaS. This PR adds the Makefile rules and terraform configuration to enable developers to run the restore process during a disaster recovery.

### Trello card 

https://trello.com/c/yVw2bqAe

### CODE REVIEW

**Comment by merrilmathew** [comment]
[Understand and document backup restore plan for PaaS databases](https://trello.com/c/yVw2bqAe/357-understand-and-document-backup-restore-plan-for-paas-databases)

File: docs/restore-paas-postgres-database.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+The username of the person being added to the role can be obtained by asking them to run `cf target`, this is usually their @digital.education.gov.uk email address unless they have enabled Google SSO in their GovUK PaaS account.
+
+## Restoring from a point in time backup
+
+### 1. Set the correct space context and stop any apps bound to the database service instance
+
+Set the context for the cf cli commands to the space under which the service exists.
+
+Stop the app that depend on the database service instance using the `cf stop` command. The app name can be obtained by running `cf apps`.
```
Comment (NickGraham101) [inline docs/restore-paas-postgres-database.md:26]: app or apps plural?

File: docs/restore-paas-postgres-database.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+- Makefile
+
+`make <ENV> set-space-developer USER_ID=<user id obtained from $(cf target) command> `
+
+The username of the person being added to the role can be obtained by asking them to run `cf target`, this is usually their @digital.education.gov.uk email address unless they have enabled Google SSO in their GovUK PaaS account.
+
+## Restoring from a point in time backup
+
+### 1. Set the correct space context and stop any apps bound to the database service instance
```
Comment (NickGraham101) [inline docs/restore-paas-postgres-database.md:22]: Is there an existing process for putting up a maintenance page?  Might be useful to link to that here

File: terraform/dev.backend.tfvars [hunk 3]
```diff
@@ -1,2 +1,3 @@
 storage_account_name = "s165d01tfstate"
 key                  = "dev.tfstate"
+resource_group_name  = "s165d01-rg"
```
Comment (NickGraham101) [inline terraform/dev.backend.tfvars:3]: Does this property need adding to the `backend.tfvars` files for other environments?

File: Makefile [hunk 4]
```diff
@@ -45,6 +50,44 @@ validate-infra-secrets: read-keyvault-config install-fetch-config set-azure-acco
 	bin/fetch_config.rb -s azure-key-vault-secret:${KEY_VAULT_NAME}/${KEY_VAULT_SECRET_NAME} -d quiet \
 		&& echo Data in ${KEY_VAULT_NAME}/${KEY_VAULT_SECRET_NAME} looks valid
 
+.PHONY: set-space-developer
+set-space-developer: read-deployment-config ## `make dev set-space-developer USER_ID=first.last@digital.education.gov.uk`
+	$(if $(USER_ID), , $(error Missing environment variable "USER_ID", USER_ID required for this command to run))
+	cf set-space-role $(USER_ID) dfe $(SPACE) SpaceDeveloper
+
+.PHONY: unset-space-developer
+unset-space-developer: read-deployment-config ## `make dev unset-space-developer USER_ID=first.last@digital.education.gov.uk`
```
Comment (NickGraham101) [inline Makefile:59]: Would be cleaner to put comments on their own line rather than in the middle of the commands

File: Makefile [hunk 5]
```diff
@@ -45,6 +50,44 @@ validate-infra-secrets: read-keyvault-config install-fetch-config set-azure-acco
 	bin/fetch_config.rb -s azure-key-vault-secret:${KEY_VAULT_NAME}/${KEY_VAULT_SECRET_NAME} -d quiet \
 		&& echo Data in ${KEY_VAULT_NAME}/${KEY_VAULT_SECRET_NAME} looks valid
 
+.PHONY: set-space-developer
+set-space-developer: read-deployment-config ## `make dev set-space-developer USER_ID=first.last@digital.education.gov.uk`
```
Comment (NickGraham101) [inline Makefile:54]: Would be cleaner to put comments on their own line rather than in the middle of the commands

Comment (saliceti) [reply Makefile:54]: I think this allows generating help, like in apply:

```
% make
help                           Show this help
build                          Create a new image
setup                          Set up a clean database and node_modules folder for running the app or the specs in docker
stop                           bring down the containers
test                           Run the linters and specs
...
```

File: Makefile [hunk 6]
```diff
@@ -45,6 +50,44 @@ validate-infra-secrets: read-keyvault-config install-fetch-config set-azure-acco
...
+	$(if $(USER_ID), , $(error Missing environment variable "USER_ID", USER_ID required for this command to run))
+	cf set-space-role $(USER_ID) dfe $(SPACE) SpaceDeveloper
+
+.PHONY: unset-space-developer
+unset-space-developer: read-deployment-config ## `make dev unset-space-developer USER_ID=first.last@digital.education.gov.uk`
+	$(if $(USER_ID), , $(error Missing environment variable "USER_ID", USER_ID required for this command to run))
+	cf unset-space-role $(USER_ID) dfe $(SPACE) SpaceDeveloper
```
Comment (NickGraham101) [inline Makefile:61]: Should be `${}` as used elsewhere in this file rather than `$()`

File: Makefile (hunk 5 above)
Comment (NickGraham101) [reply Makefile:54]: That doesn't work in this repo (not sure why) :(
The syntax is also slightly different, to match apply it should be:
```
.PHONY: set-space-developer
set-space-developer:  ## `make dev set-space-developer USER_ID=first.last@digital.education.gov.uk`
   read-deployment-config
   $(if $(USER_ID), , $(error Missing environment variable "USER_ID", USER_ID required for this command to run))
   cf set-space-role $(USER_ID) dfe $(SPACE) SpaceDeveloper
```
The script that's being executed is on the line after the comments in apply

File: Makefile [hunk 7]
```diff
@@ -45,6 +50,44 @@ validate-infra-secrets: read-keyvault-config install-fetch-config set-azure-acco
...
+rename-postgres-service: read-deployment-config ## `make dev rename-postgres-service NEW_NAME_SUFFIX=old CONFIRM_RENAME`
+	$(if $(CONFIRM_RENAME), , $(error can only run with CONFIRM_RENAME))
+	$(if $(NEW_NAME_SUFFIX), , $(error NEW_NAME_SUFFIX is required))
+	cf target -s ${SPACE} > /dev/null
+	cf rename-service  ${POSTGRES_DATABASE_NAME} ${POSTGRES_DATABASE_NAME}-$(NEW_NAME_SUFFIX)
+
+remove-postgres-tf-state: terraform-init ## `make dev remove-postgres-tf-state PASSCODE=XXX`
+	cd terraform && terraform state rm cloudfoundry_service_instance.postgres
+
+restore-postgres: terraform-init read-deployment-config ## `make dev restore-postgres DB_INSTANCE_GUID="<cf service db-name --guid>" BEFORE_TIME="" TF_VAR_api_docker_image=ghcr.io/dfe-digital/qualified-teachers-api:<COMMIT_SHA> PASSCODE=<auth code from https://login.london.cloud.service.gov.uk/passcode>`
```
Comment (NickGraham101) [inline Makefile:82]: Can the time format be included, eg `BEFORE_TIME="yyyy-MM-dd hh:mm:ss"` (or whatever the correct format is)

File: docs/restore-paas-postgres-database.md (hunk 2 above)
Comment (merrilmathew) [reply docs/restore-paas-postgres-database.md:22]: No maintenance page for the API. /status will return error if any dependencies (postgres, Redis, CRM) are down

File: terraform/dev.backend.tfvars (hunk 3 above)
Comment (merrilmathew) [reply terraform/dev.backend.tfvars:3]: Yes this needs adding to all environment to run terraform commands locally

File: Makefile (hunk 5 above)
Comment (merrilmathew) [reply Makefile:54]: Help now works after including help rule :)

**Review by NickGraham101 — APPROVED** [review]

**Pull request merged**